### PR TITLE
Use Listwise loss function for LTR

### PR DESCRIPTION
**Pull request by bilbof**
This changes the loss function to LIST_MLE_LOSS [1].

This loss function leads to a better NDCG score (about 1% better).

From Wikipedia [2]:
> In practice, listwise approaches often outperform pairwise approaches

Demonstration of improvement (red line is with this new loss function) using a month of training data:

Screenshot 2019-12-02 at 11 56 50

[1] https://github.com/tensorflow/ranking/blob/v0.2.0/tensorflow_ranking/python/losses.py#L44
[2] https://en.wikipedia.org/wiki/Learning_to_rank#Approaches

More details on the loss function in the paper that released it: http://icml2008.cs.helsinki.fi/papers/167.pdf

https://trello.com/c/XGxd6mNO

### CODE REVIEW

File: ltr_scripts/tf_ranking_libsvm.py [hunk 1]
```diff
@@ -356,6 +356,7 @@ def get_eval_metric_fns():
   metric_fns.update({
       "metric/%s" % name: tfr.metrics.make_ranking_metric_fn(name) for name in [
           tfr.metrics.RankingMetricKey.ARP,
+          tfr.metrics.RankingMetricKey.MRR, # Mean Reciprocal Rank
```
Comment (sihugh) [inline ltr_scripts/tf_ranking_libsvm.py:359]: Where is this used?

**Review by sihugh — COMMENTED** [review]
Graphs look great!  Just a couple of questions about the context 😁

File: ltr_scripts/tf_ranking_libsvm.py (hunk 1 above)
Comment (bilbof) [reply ltr_scripts/tf_ranking_libsvm.py:359]: This reports mean reciprocal rank as the model is trained, so we can visualise the effect of training in tensorboard (it doesn't impact training as far as I am aware, it's just logging metrics):

Screenshot 2019-12-02 at 16 04 59

File: ltr_scripts/tf_ranking_libsvm.py [hunk 2]
```diff
@@ -81,7 +81,7 @@
 flags.DEFINE_integer("train_batch_size", 32, "The batch size for training.")
 flags.DEFINE_integer("num_train_steps", 100000, "Number of steps for training.")
 
-flags.DEFINE_float("learning_rate", 0.01, "Learning rate for optimizer.")
+flags.DEFINE_float("learning_rate", 0.1, "Learning rate for optimizer.")
```
[the remark this answers is not in the document]
Comment (bilbof) [reply ltr_scripts/tf_ranking_libsvm.py:84]: The learning rate determines how fast the internal weights of the model change. I picked it as it's possible to train the model faster if you use a bigger learning rate. The optimizer we use ([Adagrad](https://databricks.com/glossary/adagrad)) modifies the learning rate over time. I understood this as it could start fast and become slower/faster automatically later on, but I could be wrong here!

[the remark this answers is not in the document]
Comment (sihugh) [reply ltr_scripts/tf_ranking_libsvm.py:84]: Cool, so we might have to keep an eye on it, or try some alternatives at some point?

**Review by sihugh — APPROVED** [review]

File: ltr_scripts/tf_ranking_libsvm.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (bilbof) [reply ltr_scripts/tf_ranking_libsvm.py:84]: Yes exactly. I only tried `0.1`; we could try others like `1, 0.0001`.

File: ltr_scripts/train.sh [hunk 3]
```diff
@@ -4,7 +4,7 @@ OUTPUT_DIR=${OUTPUT_DIR:-../tmp/libsvm}
 TRAIN=${TRAIN:-../tmp/ltr_data/train.txt}
 VALI=${VALI:-../tmp/ltr_data/validate.txt}
 TEST=${TEST:-../tmp/ltr_data/test.txt}
-STEPS=${STEPS:-50000} # runs at ~1,000 rounds per minute. 50,000 steps = 50 mins.
+STEPS=${STEPS:-100000} # runs at ~1,000 rounds per minute. 50,000 steps = 50 mins.
```
Comment (barrucadu) [inline ltr_scripts/train.sh:7]: Should this change have been committed? (and if it should can the comment be updated?)
